src/storage/gc_worker.rs

```diff
@@ -408,6 +448,550 @@ impl<E: Engine> Runnable<GCTask> for GCRunner<E> {
+        let mut progress = Some(Key::from_encoded(BEGIN_KEY.to_vec()));
+
+        // Records last time checking safe point from PD.
+        let mut timer = Instant::now();
+        // Records how many region we have GC-ed.
+        let mut processed_regions = 0;
+
+        info!(
+            "gc_worker: started auto gc with safe point {}",
+            self.safe_point
+        );
+
+        loop {
+            self.gc_worker_ctx.check_stopped()?;
+
+            if need_rewind {
+                if progress.is_none() {
+                    // Worked to the end. restart from beginning
+                    progress = Some(Key::from_encoded(BEGIN_KEY.to_vec()));
+                    need_rewind = false;
+                    info!(
+                        "gc_worker: auto gc has rewound after gc {} regions",
+                        processed_regions
+                    );
+                    processed_regions = 0;
+                }
+            } else if progress.is_none()
+                || (end.is_some() && progress.as_ref().unwrap() >= end.as_ref().unwrap())
+            {
+                info!(
+                    "gc_worker: finished auto gc, {} regions processed",
+                    processed_regions
+                );
+                return Ok(());
+            }
+
+            assert!(progress.is_some());
+
+            // Update safe_point periodically
+            if self.need_update_safe_point(&timer) {
+                if self.try_update_safe_point() {
+                    if progress.as_ref().unwrap().as_encoded().is_empty() {
+                        // `progress` is empty means the starting. We don't need to rewind. We just
+                        // Continue GC to the end.
+                        need_rewind = false;
+                        end = None;
+                    } else {
+                        need_rewind = true;
+                        end = progress.clone();
+                        info!(
+                            "gc_worker: safe point updated to {}, auto gc will rewind to {}",
+                            self.safe_point,
+                            end.as_ref().unwrap()
+                        );
+                    }
+                }
+                timer = Instant::now();
+            }
+
+            progress = self.gc_once(progress.unwrap(), &mut processed_regions)?;
+        }
+    }
+
+    /// Checks whether we need to check safe point now.
+    #[inline]
+    fn need_update_safe_point(&self, last_update_time: &Instant) -> bool {
+        last_update_time.elapsed() >= self.cfg.poll_safe_point_interval
+            || self.cfg.always_check_safe_point
+    }
+
+    /// GC the next region after `from_key`. Returns the end key of the region it processed.
+    /// If we have processed to the end of all regions, returns `None`.
+    fn gc_once(
+        &mut self,
+        from_key: Key,
+        processed_regions: &mut usize,
+    ) -> GCManagerResult<Option<Key>> {
```

Comment:
Seems the return type don't need `GCManagerResult` to wrap.

Reply:
It can be broken before it finishing, so use `GCManagerResult` here.